src/MPAGSCipher/PlayfairCipher.cpp

```diff
@@ -0,0 +1,171 @@
+    
+    // Remove duplicated letters
+    auto iter2 = std::remove_if(key_.begin(),key_.end(),[=](char& val){
+        if(encounteredletters.find(val)!=std::string::npos){
+            return true;
+        } else {
+            encounteredletters += val; 
+            return false;
+        }});
+    key_.erase(iter2,key_.end());
+
+    // Store the coords of each letter and the playfair cipher key map
+
+    for(std::size_t i{0};i<key_.size();i++){
+        Str2CoordMap::value_type invcoords{std::pair<int,int> {(i/5)+1,(i%5)+1},key_[i]};
+        Str2CharMap::value_type coords{key_[i],std::pair<int,int> {(i/5)+1,(i%5)+1}};
+
+        playfairciphermap_.insert(coords);
+        playfairciphermapinv_.insert(invcoords);
+    }
+    /*
+    for ( auto p : playfairciphermap_ ){
+        std::cout << p.first << ": " << p.second.first << ":" << p.second.second << std::endl;
+    }
+
+    for (auto p : playfairciphermapinv_){
+        std::cout << p.first.first << ":" << p.first.second << " :" << p.second << std::endl;
+    }
+    */
+    
+}
+
+
+std::string PlayfairCipher::applyCipher(const std::string& inputText,
+                                        const CipherMode cipherMode) const {
+
+    std::string output{""},modifiedinput{""},input{inputText};
+
+    switch(cipherMode){
+        case CipherMode::Encrypt:
+            // Change J → I
+            std::transform(input.begin(),input.end(),input.begin(),[](char& val){if(val=='J'){return 'I';} else {return val;}});
+            std::cout << "after j->i: " << input << std::endl;
+            // If repeated chars in a digraph add an X or Q if XX
+            modifiedinput += input[0];
+            for(std::size_t j{1};j<input.size();j++){
+                if(input[j] == input[j-1]){
+                    if(input[j-1] != 'X'){
+                        modifiedinput += 'X';
+                        modifiedinput += input[j];
+                    } else {
+                        modifiedinput += 'Q';
+                        modifiedinput += input[j];
+                    }
+                } else {
+                    modifiedinput += input[j];
+                }
+            }
+            std::cout << "after double char correction: " << modifiedinput << std::endl;
+            // if the size of input is odd, add a trailing Z
+            if(modifiedinput.size()%2 != 0){
+                modifiedinput += 'Z';
+            }
+            std::cout << "after Z: " << modifiedinput << std::endl;
+            // Loop over the input in Digraphs
+            for(std::size_t k{0};k<modifiedinput.size();k+=2){
+                std::cout << "iteration var: " << k  << "chars: " << modifiedinput[k] << "+" << modifiedinput[k+1] << std::endl;
+                char digraphchar1{modifiedinput[k]};
+                char digraphchar2{modifiedinput[k+1]};
+
+                // - Find the coords in the grid for each digraph
+
+                auto iterindex1{playfairciphermap_.find(digraphchar1)};
```

Comment:
You can just use `[]` or (even better) `at` here as then you will be returned the `pair` rather than an iterator. e.g., this becomes:
```
std::pair<int,int> charcoord1{playfairciphermap_.at(digraphchar1).second;
```